arbos/l1pricing/l1pricing.go

```diff
@@ -7,239 +7,365 @@ import (
-	return ps.userSpecifiedAggregatorsForAddress(sender).GetAnyMember()
+func (ps *L1PricingState) SetUnitsSinceUpdate(units uint64) error {
+	return ps.unitsSinceUpdate.Set(units)
 }
 
-func (ps *L1PricingState) SetUserSpecifiedAggregator(sender common.Address, maybeAggregator *common.Address) error {
-	paSet := ps.userSpecifiedAggregatorsForAddress(sender)
-	if err := paSet.Clear(); err != nil {
+func (ps *L1PricingState) AddToUnitsSinceUpdate(units uint64) error {
+	oldUnits, err := ps.unitsSinceUpdate.Get()
+	if err != nil {
 		return err
 	}
-	if maybeAggregator == nil {
-		return nil
-	}
-	return paSet.Add(*maybeAggregator)
+	return ps.unitsSinceUpdate.Set(oldUnits + units)
+}
+
+func (ps *L1PricingState) PricePerUnit() (*big.Int, error) {
+	return ps.pricePerUnit.Get()
+}
+
+func (ps *L1PricingState) SetPricePerUnit(price *big.Int) error {
+	return ps.pricePerUnit.Set(price)
 }
 
-func (ps *L1PricingState) RefusesDefaultAggregator(addr common.Address) (bool, error) {
-	val, err := ps.refuseDefaultAggregator.Get(common.BytesToHash(addr.Bytes()))
+// Update the pricing model based on a payment by a batch poster
+func (ps *L1PricingState) UpdateForBatchPosterSpending(statedb vm.StateDB, evm *vm.EVM, updateTime uint64, currentTime uint64, batchPoster common.Address, weiSpent *big.Int) error {
+	batchPosterTable := ps.BatchPosterTable()
+	posterState, err := batchPosterTable.OpenPoster(batchPoster)
 	if err != nil {
-		return false, err
+		return err
 	}
-	return val != (common.Hash{}), nil
-}
 
-func (ps *L1PricingState) SetRefusesDefaultAggregator(addr common.Address, refuses bool) error {
-	val := uint64(0)
-	if refuses {
-		val = 1
+	// compute previous shortfall
+	totalFundsDue, err := batchPosterTable.TotalFundsDue()
+	if err != nil {
+		return err
 	}
-	return ps.refuseDefaultAggregator.Set(common.BytesToHash(addr.Bytes()), common.BigToHash(arbmath.UintToBig(val)))
-}
+	fundsDueForRewards, err := ps.FundsDueForRewards()
+	if err != nil {
+		return err
+	}
+	oldShortfall := am.BigSub(statedb.GetBalance(L1PricerFundsPoolAddress), am.BigAdd(totalFundsDue, fundsDueForRewards))
 
-// Get the aggregator who is eligible to be reimbursed for L1 costs of txs from sender, or nil if there is none.
-func (ps *L1PricingState) ReimbursableAggregatorForSender(sender common.Address) (*common.Address, error) {
-	fromTable, err := ps.UserSpecifiedAggregator(sender)
+	// compute allocation fraction -- will allocate updateTimeDelta/timeDelta fraction of units and funds to this update
+	lastUpdateTime, err := ps.LastUpdateTime()
 	if err != nil {
-		return nil, err
+		return err
+	}
+	if lastUpdateTime == 0 && currentTime > 0 { // it's the first update, so there isn't a last update time
+		lastUpdateTime = updateTime - 1
 	}
-	if fromTable != nil {
-		return fromTable, nil
+	if updateTime > currentTime || updateTime < lastUpdateTime || currentTime == lastUpdateTime {
```

Comment:
This could be simplified to `updateTime >= currentTime || updateTime < lastUpdateTime`

Reply:
Done